Week3 [STEP 1] Hisop

Hi-sop opened the pull request:
@juun97 
STEP1 PR드립니다.

## 고민했던 점

class와 struct중 고민하다 상속을 사용하지 않아도 된다 생각하여 Struct를 선택했습니다.
Person과 CoffeShop모두 자신의 프로퍼티(money, sales,pickUpTable)를 수정해야했으므로 mutating 키워드를 사용했습니다.

espresso는 커피 종류에 포함되지만 만든 카페에서 메뉴로 판매하진 않는다는 컨셉으로 넣어보았습니다.

Commit:
Feat: Step1 요구사항 구현
class와 struct중 고민. 상속을 사용하지 않아서 Struct 선택
구매와 주문받기 커피만들기 메서드 생성
정확한 지시가 없는 항목은 임의대로 설정

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+    case latte = 4500
+    case cappuccino = 5000
+    case espresso = 3500
+}
```
enum 타입의 case 와 rawValue 를 잘 활용 해주셨네요 👍🏻

하지만 좀 더 생각해봐야할 점이 있습니다.

1.  raw value 타입명시의 컨벤션 수정이 필요해보이네요 🤔
2. Coffee 의 가격같은 경우 CoffeeShop 마다 다른 가격으로 존재할 수도 있는데 해당 타입에서 가격을 가지게 한 이유가 있을까요?
3. 추가적으로, raw value 를 Int 로 설정하고 값을 따로 설정하지 않을 경우 어떻게 되는지 아시나요?

Hi-sop replied:
일단 커피숍에서 할당하는것으로 변경했으며 추후 수정이 필요하거나 카페마다 다른 메뉴가 필요하면 메뉴추가, 메뉴수정 등의 함수를 구현해 해결할 것 같습니다.

값을 따로 설정하지 않으면 0, 1, 2, 3 같은 방식으로 할당됩니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+    var sales = 0
+    var barista: Person?
+    var menu: [Coffee] = []
+    var pickUpTable: [Coffee] = []
```
1. Hisop 님의 Type Annotation 과 Type Reference 를 사용하는 기준이 있을까요?

프로퍼티의 Type 컨벤션이 일정하지가 않습니다 🤔

2. CoffeeShop의 요구사항 명세를 보면 `메뉴판(커피 종류, 가격)을 가질 수 있도록 해봅시다.` 라는 항목이 있습니다. CoffeeShop 에서 Coffee 가격을 관리할 생각도 해보셨을까요?

Hi-sop replied:
추론으로 생성이 된다면 추론을 사용했습니다 -> 전부 명시하도록 수정했습니다.

enum배열로 관리하던 menu를 Dictionary 타입으로 변경하고 카페에서 가격을 생성하는것으로 변경했습니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+    var menu: [Coffee] = []
+    var pickUpTable: [Coffee] = []
+    
+    mutating func takeOrder(_ coffee: Coffee, from: Person) -> Bool {
```
네이밍을 통해 해당 메서드가 어떤 기능을 담당하는지 알 수 있어서 좋았습니다 👍🏻
또한, Bool 타입의 반환타입을 사용해 성공여부를 전달해주는 점이 인상깊네요 😄

하지만 전달인자 레이블의 활용이 어색해 보입니다.
사용한 시점에서 전달인자 레이블을 통해 문맥의 흐름을 맞춰 주는점은 매우 좋으나 현재 상황에서는 전달인자 레이블을 사용하지 않고 파라미터의 네이밍이 from 이 되어 함수 내부에서 네이밍의 어색함이 느껴지네요 🤔

Hi-sop replied:
전달인자 레이블의 이름을 수정했습니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+        if from.money < coffee.rawValue {
+            print("잔액이 \(from.money - coffee.rawValue)모자랍니다.")
+            return false
+        }
```
1. takeOrder 네이밍이 주문을 받는다는 기능을 의도하신 거 같은데 내부에서 주문 받은 메뉴에 대한 검증도 하고 있네요 🤔

기능에 맞게 분리를 해볼까요 😄

2. 위에선 guard 문을 사용하고 아래에선 if 문을 사용해주신 이유가 있으신가요?

Hi-sop replied:
주문 검증 함수를 따로 분리했습니다.

두 코드블럭을 다른 시점에 만들어서 신경쓰지 못한 듯 합니다. 빠른 종료를 위한 케이스이니 guard로 변경했습니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+            return false
+        }
+        sales = sales + coffee.rawValue
+        self.makeCoffee(coffee, from: from.name)
```
self 키워드를 명시적으로 사용하신 이유가 있으실까요?
추가적으로 self 와 Self 의 차이점에 대해 알고계시나요?

Hi-sop replied:
C++문법을 쓰던 버릇 같습니다. 혼동되는 네이밍이 없으므로 self를 쓰지 않는것으로 수정했습니다.

찾아보니 Self는 타입을 가리킨다고 하네요!

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+        return true
+    }
+    
+    mutating func makeCoffee(_ coffee: Coffee, from name: String) {
+        pickUpTable.append(coffee)
+        print("\(name) 님이 주문하신 \(coffee)(이/가) 준비되었습니다. 픽업대에서 가져가주세요.")
+    }
```
mutating 키워드는 어떤 상황에서 사용을 하는걸까요?

Hi-sop replied:
구조체내의 메소드에서 프로퍼티의 값을 수정해야하는 경우 붙이는 키워드입니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,58 @@
+        print("\(name) 님이 주문하신 \(coffee)(이/가) 준비되었습니다. 픽업대에서 가져가주세요.")
+    }
+    
+    init() {
+        menu.append(Coffee.americano)
+        menu.append(Coffee.latte)
+        menu.append(Coffee.cappuccino)
+    }
```
init() 의 역할에 대해서 알고 계시나요?

시간이 되시면 class의 init 과 struct 의 init 에 대해서 비교해보시겠어요?

Hi-sop replied:
class는 init을 명시하지 않으면 오류가 발생합니다.
구조체는 memberwise 기능을 제공해 오류가 발생하지 않습니다.

이유는 상속의 가능성이라고 합니다. [해당 페이지](https://jayb-log.tistory.com/258) 를 참고했습니다.

juun97 reviewed (commented):
## 고민했던 점
> class와 struct중 고민하다 상속을 사용하지 않아도 된다 생각하여 Struct를 선택했습니다.
Person과 CoffeShop모두 자신의 프로퍼티(money, sales,pickUpTable)를 수정해야했으므로 mutating 키워드를 사용했습니다.
espresso는 커피 종류에 포함되지만 만든 카페에서 메뉴로 판매하진 않는다는 컨셉으로 넣어보았습니다.

class 와 struct 를 고민하는것은 앞으로도 많이 접하셔야 하기에 매우 중요한 주제입니다.
그렇기에 제가 몇가지 숙제를 드리려고 합니다.
1. 어떤 상황에서 class와 struct 를 사용하는지 알아오기
2. Apple 에선 어떤것을 권장하는지

Hisop님도 잘 아시다시피 class 는 reference Type struct는 Value 타입입니다. 

현재 Coffee shop 을 struct 로 구현해주셨기에 Person의 커피주문이 Coffee Shop의 변경을 필요로 하기에 inout 파라미터를 사용하게 되신걸로 보입니다.

inout 파라미터가 물론 사용을 하라고 존재한 키워드 이지만 분명히 확실한 단점이 존재합니다.

1. Side Effects 증가

inout 파라미터를 사용하면 함수 내부에서 해당 변수의 값을 변경할 수 있습니다. 이는 함수 외부에서 예상치 못한 결과를 초래할 수 있으며, 함수가 예상치 못한 부작용을 가질 수 있습니다. 함수가 여러 곳에서 호출되고 inout 파라미터를 공유하는 경우 코드의 복잡성이 증가하고 버그를 유발할 수 있습니다.

2. 복잡성 증가

inout 파라미터를 사용하면 함수가 변수의 메모리 주소를 전달받아 값을 수정하므로, 일반적인 함수보다 더 복잡해질 수 있습니다. 이로 인해 코드의 가독성이 떨어질 수 있고, 디버깅이 어려워질 수 있습니다.

---

```swift
struct TypeA {
    //...
}

class TypeB {
    //...
}

let instanceA = TypeA()
let instanceB = instanceA

let instanceC = TypeB()
let instanceD = instanceC
```

추가적으로 Hisop님은 인스턴스의 개념에 대해서 알고 계시나요?
다음과 같은 Type들이 있고 Instance가 존재할 때 

1. instanceA 와 instanceB 는 같은 인스턴스 일까요?
2. instanceC 와 instanceD 는 같은 인스턴스 일까요?

한 번 찾아보시고 제가 드린 질문에 대해서 답변을 주시고, 그 답변이 나오기 까지 어떤 생각을 하셨는지 알려주시겠어요?

매우 중요한 부분이라 숙제가 좀 있습니다 😄

Commit:
Refactor: 피드백 받은 내용 수정
커피숍 생성시 커피 가격 설정하는것으로 변경
타입 추론으로 생성한 프로퍼티 타입 명시로 변경
guard, if 혼용되던것 guard로 변경
self붙여 호출하던 함수들 self제거
menu Coffee배열에서 Dictionary 타입으로 변경

Hi-sop commented:
애플 공식문서에선 기본적으로 Struct를 사용. 상속을 사용하거나 Objective-C 코드를 사용할 경우 Class를 사용하라고 권장하네요. [해당 페이지](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes) 참고했습니다.

1. instanceA 와 instanceB 는 같은 인스턴스 일까요?
    다른 인스턴스입니다. 구조체이므로 깊은 복사를 시행하며 값을 복사해도 복사본만 변경됩니다.

2. instanceC 와 instanceD 는 같은 인스턴스 일까요?
    다른 인스턴스입니다. 클래스이므로 얕은 복사를 시행합니다. 값을 복사하면 원본도 같이 변경됩니다. D는 C와 같은 주소를 바라봅니다

juun97 commented:
Struct 와 Class 의 차이를 잘 알고 게시군요 👍🏻
수정사항도 깔끔하게 잘 반영이 되었습니다.

다만, 현재는 모든 수정사항이 하나의 커밋으로 올라와있습니다. 작성한 기능 단위나 Hisop 님만의 기준에 따라 커밋도 분리해서 해보시는 걸 권장드립니다 😃

일단 머지는 할테니 다음 스텝부터 커밋도 한번 관리해볼까요?
